### PR TITLE
[FLINK-5067] Make Flink compile with 1.8 Java compiler

### PR DESCRIPTION
**Pull request by melentye**
- New type inference rules changed which overloaded methods are picked.
- Renamed java.version property in maven to java.minor.version. This is due to the fact that overriding java.version is used by some third-party classes and they expect the version to be fully qualified.
- Introduced Java 8 build for Travis CI

### CODE REVIEW

File: flink-tests/src/test/java/org/apache/flink/test/windowing/sessionwindows/SessionWindowITCase.java [hunk 1]
```diff
@@ -122,11 +122,11 @@ private void runTest(
...
 		// check that overall event counts match with our expectations. remember that late events within lateness will
 		// each trigger a window!
-		Assert.assertEquals(
-				(LATE_EVENTS_PER_SESSION + 1) * NUMBER_OF_SESSIONS * EVENTS_PER_SESSION,
+		Assert.assertEquals(Long.valueOf(
+				(LATE_EVENTS_PER_SESSION + 1) * NUMBER_OF_SESSIONS * EVENTS_PER_SESSION),
 				result.getAccumulatorResult(SESSION_COUNTER_ON_TIME_KEY));
-		Assert.assertEquals(
-				NUMBER_OF_SESSIONS * (LATE_EVENTS_PER_SESSION * (LATE_EVENTS_PER_SESSION + 1) / 2),
+		Assert.assertEquals(Long.valueOf(
```
Comment (StephanEwen) [inline flink-tests/src/test/java/org/apache/flink/test/windowing/sessionwindows/SessionWindowITCase.java]: I think unboxing the second argument would be nicer than boxing the first

File: flink-java/src/test/java/org/apache/flink/api/common/operators/CollectionExecutionAccumulatorsTest.java [hunk 2]
```diff
@@ -35,7 +35,7 @@
 	@Test
 	public void testAccumulator() {
 		try {
-			final int NUM_ELEMENTS = 100;
+			final Integer NUM_ELEMENTS = 100;
```
Comment (StephanEwen) [inline flink-java/src/test/java/org/apache/flink/api/common/operators/CollectionExecutionAccumulatorsTest.java]: Can we stay, whereever possible, with primitive types? Rather unbox the type at a different place?

File: pom.xml [hunk 3]
```diff
@@ -96,7 +96,7 @@ under the License.
 		<slf4j.version>1.7.7</slf4j.version>
 		<guava.version>18.0</guava.version>
 		<akka.version>2.3.7</akka.version>
-		<java.version>1.7</java.version>
+		<java.minor.version>1.7</java.minor.version>
```
Comment (StephanEwen) [inline pom.xml]: Does this make a difference, or is that mainly a naming preference?
It it is only about naming, I would prefer to undo this change.
Call it "being conservative" with `pom.xml` changes - they have subtle implication son release scripts, etc.

File: flink-java/src/test/java/org/apache/flink/api/common/operators/CollectionExecutionAccumulatorsTest.java (hunk 2 above)
Comment (melentye) [inline flink-java/src/test/java/org/apache/flink/api/common/operators/CollectionExecutionAccumulatorsTest.java]: Absolutely!

File: flink-tests/src/test/java/org/apache/flink/test/windowing/sessionwindows/SessionWindowITCase.java (hunk 1 above)
Comment (melentye) [inline flink-tests/src/test/java/org/apache/flink/test/windowing/sessionwindows/SessionWindowITCase.java]: Sure thing, fixed

**Comment by melentye** [comment]
@StephanEwen I initially introduced a new Travis profile so that the coverage would be strictly improved, i.e. all old combinations tests plus the new ones. But I am also fine reusing some of the existing profiles too, PR updated.

File: pom.xml (hunk 3 above)
Comment (melentye) [inline pom.xml]: I actually prefer java.version for it's conciseness and I share the conservative tune in regards to unnecessary pom.xml changes. But in this case I did it on purpose, long explanation follows.

The java.version is passed by -Djava.version=1.8 argument to mvn which seems to be Maven-recommended (or even the only one supported) way of redefining the property from command-line. An inconvenient side-effect is that this property is then propagated down until the JVM that executes the tests and causes java.lang.RuntimeException: Unexpected version format: 1.8	at org.apache.hadoop.hbase.util.ClassSize.<clinit>(ClassSize.java:119) starting in org.apache.flink.addons.hbase.TableInputFormatITCase (won't bore you with the complete stack trace). Since this is third party code, we can't really fix it in Flink source. Therefore I thought to rename the property instead to avoid the clash.

There's a way to keep the java.version property name as is: we can use a maven profile to set java.version. In this case when you activate the profile with -P argument to mvn, the java.version will not be propagated as a system property to the JVM running the tests and won't cause problems.

I've noticed that there's an existing profile called "jdk8" which sounds relevant. But it's currently automatically activated if you're running Maven with Java 8 so that could turn out to be a surprise for the maintainers to realize that they suddenly are building 1.8 target bytecode. The profile doesn't do much though so I am questioning if it really needs to be auto-activated. I think it'll be cleaner to stop auto-activating it and reuse it for the above purpose. See proposal in the PR update.

**Comment by StephanEwen** [comment]
Looks good!

Just curious, you mentioned that using the `java.version` variable clashes with how it is used by the JVM. Is that not an issue any more?

**Comment by melentye** [comment]
@StephanEwen it seems that the semantics is different when java.version is just defined in pom.xml versus being overriden with -Djava.version=x.y.z argument passed to mvn.

org.apache.hadoop.hbase.util.ClassSize expects a very specific format of java.version system property. But when java.version is only defined in pom.xml then it doesn't actually become a system property unlike in case of using -Djava.version

**Comment by rmetzger** [comment]
Looks good to merge.

I'm currently verifying the change, then, I'll merge it: https://travis-ci.org/rmetzger/flink/builds/207494237
